service/matching/physical_task_queue_manager.go

```diff
@@ -317,14 +317,28 @@ func (c *physicalTaskQueueManagerImpl) PollTask(
 		return c.matcher.PollForQuery(ctx, pollMetadata)
 	}
 
-	task, err := c.matcher.Poll(ctx, pollMetadata)
-	if err != nil {
-		return nil, err
-	}
+	for {
+		task, err := c.matcher.Poll(ctx, pollMetadata)
+		if err != nil {
+			return nil, err
+		}
+
+		// It's possible to get an expired task here: taskReader checks for expiration when
+		// reading from persistence, and physicalTaskQueueManager checks for expiration in
+		// ProcessSpooledTask, but the one task blocked in the matcher could expire while it's
+		// there. In that case, go back for another task.
+		// If we didn't do this, the task would be rejected when we call RecordXTaskStarted on
+		// history, but this is more efficient.
+		if task.event != nil && IsTaskExpired(task.event.AllocatedTaskInfo) {
+			c.taggedMetricsHandler.Counter(metrics.ExpiredTasksPerTaskQueueCounter.Name()).Record(1)
+			task.finish(nil)
```

Comment:
the comment on `task.finish` says 
```
// finish marks a task as finished. Should be called after a poller picks up a task
// and marks it as started. If the task is unable to marked as started, then this
// method should be called with a non-nil error argument.
```
Should this be called with a non-nil error saying something about task expiration? Maybe that is not necessary because we're dropping the expired task here, not picking it up. It seems like only the `task.event.completionFunc` part of `task.finish` will matter, so, do we want to make sure that `task.event.completionFunc` is called with an error that explains that the task was not actually started?

Reply:
https://github.com/temporalio/temporal/blob/b702f2ab96e8619f136c894576e434ec0c19e499/service/matching/task_reader.go#L274 this use of `IsTaskExpired` does not call `task.finish`, for example

Reply:
`nil` means the task does not have be handled anymore, so it's appropriate here (there's nothing more to do if it's expired). If it was not nil, it would try to add the task back again: https://github.com/temporalio/temporal/blob/main/service/matching/backlog_manager.go#L233-L244

What we really want from finish here is for backlogManager.completeTask to call ackManager.completeTask (line 263) so it knows it can move the ack level past this task.

For that call in taskReader, the task hasn't been given to ackManager yet (that's in addSingleTaskToBuffer), so there's no need to call finish, we can just drop it.